core/domain/exp_domain.py

```diff
@@ -3068,81 +3181,1637 @@ def _convert_states_v51_dict_to_v52_dict(
+                        ) and cls._is_enclosed_by(range_var, range_ele)
+                    ):
+                        invalid_rules.append(rule_spec)
+
+                for den in matched_denominator_list:
+                    if (
+                        rule_spec['rule_type'] ==
+                        'HasFractionalPartExactlyEqualTo'
+                    ):
+                        rule_spec_f = rule_spec['inputs']['f']
+                        assert isinstance(rule_spec_f, dict)
+                        if den['denominator'] == rule_spec_f['denominator']:
+                            invalid_rules.append(rule_spec)
+
+                ranges.append(range_var)
+                matched_denominator_list.append(matched_denominator)
+
+        empty_ans_groups = []
+        for invalid_rule in invalid_rules:
+            for answer_group in answer_groups:
+                for rule_spec in answer_group['rule_specs']:
+                    if rule_spec == invalid_rule:
+                        answer_group['rule_specs'].remove(rule_spec)
+
+                if (
+                    len(answer_group['rule_specs']) == 0 and
+                    answer_group not in empty_ans_groups
+                ):
+                    empty_ans_groups.append(answer_group)
+
+        for empty_ans_group in empty_ans_groups:
+            answer_groups.remove(empty_ans_group)
+
+        state_dict['interaction']['answer_groups'] = answer_groups
+
+    @classmethod
+    def _fix_multiple_choice_input_interaction(
+        cls, state_dict: state_domain.StateDict, state_name: str
+    ) -> None:
+        """Fixes MultipleChoiceInput interaction for the following cases:
+        - The rules should not be duplicate else the one with not pointing to
+        different state will be deleted
+        - No answer choice should appear in more than one rule else the
+        latter rule will be removed
+        - Answer choices should be non-empty and unique else will be fixed
+        accordingly
+
+        Args:
+            state_dict: state_domain.StateDict. The state dictionary that needs
+                to be fixed.
+            state_name: str. The name of the state.
+        """
+        answer_groups = state_dict['interaction']['answer_groups']
+
+        choices: List[state_domain.SubtitledHtmlDict] = (
+            state_dict['interaction']['customization_args'][
+                'choices']['value']
+        )
+        cls._choices_should_be_unique_and_non_empty(
+            choices,
+            answer_groups,
+            state_dict,
+            is_item_selection_interaction=False)
+
+        cls._remove_duplicate_rules_inside_answer_groups(
+            answer_groups, state_name)
+
+        state_dict['interaction']['customization_args']['choices'][
+            'value'] = choices
+        state_dict['interaction']['answer_groups'] = answer_groups
+
+    @classmethod
+    def _fix_item_selection_input_interaction(
```

Comment:
@seanlip previously mentioned that the Questions are very recent so there might not be any validation errors. Also, the validations run on save so there won't be any problem loading any questions. What do you think?
Thanks

Reply:
> Also, the validations run on save so there won't be any problem loading any questions. 

@lkbhitesh07 Just to confirm, you have tested this?

Reply:
@lkbhitesh07 Can yo confirm on this?

Reply:
Hey @DubeySandeep,
Sorry for the late response, I did check for the exploration and it was working perfectly fine, the validation only runs when we save the changes. For the question, It will be a problem because at the time of loading it uses `from_dict` method of state and that results in the calling of validation method.
I think I have to bring the previous changes back where I was only calling the validations for the exploration and not for the Question.
Thanks

Reply:
I'm also doing some changes for the validations in which I will be moving some validations to strict mode, along with that I will make the changes where the validations will only be called for the explorations.
The stack trace for the error which will help to understand the flow of calling validation in the questions - 
```
Traceback (most recent call last):
  File "/tmp/tmpQLqm_p/lib/python3.8/site-packages/webapp2.py", line 604, in dispatch
    return method(*args, **kwargs)
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/controllers/acl_decorators.py", line 91, in test_can_access
    return handler(self, *args, **kwargs)
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/controllers/reader.py", line 1651, in get
    question_services.get_questions_by_skill_ids(
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/domain/question_services.py", line 307, in get_questions_by_skill_ids
    questions_with_none = question_fetchers.get_questions_by_ids(question_ids)
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/domain/question_fetchers.py", line 105, in get_questions_by_ids
    questions.append(get_question_from_model(question_model))
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/domain/question_fetchers.py", line 139, in get_question_from_model
    state_domain.State.from_dict(versioned_question_state['state']),
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/domain/state_domain.py", line 5156, in from_dict
    content.validate()
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/domain/state_domain.py", line 3941, in validate
    html_cleaner.validate_rte_tags(self.html)
  File "/home/lkbhitesh07/Desktop/oppia/oppia/core/domain/html_cleaner.py", line 362, in validate_rte_tags
    raise utils.ValidationError(
core.utils.ValidationError: Image tag 'filepath-with-value' attribute should not be empty.

```

Reply:
@lkbhitesh07 just a note, when you are making the changes, do not pass any info about the exploration/question context into the from-dict method. The validation should still be called so we don't serve invalid questions. But we should handle have a code path when loading questions to handle the "question is invalid," case, ie deal with errors properly.

Reply:
Hey!
@seanlip, I and @vojtechjelinek was discussing about this part and he suggested to use the `try catch` block for the place where we call the validations at the time of loading so that we can still validate the data while storing. I don't think if we will be able to do that, the reason being is that when we call `from_dict` method inside the `question_fetchers.py` we are in the end using it to return the `Question` model, you can find the reference [here](https://github.com/oppia/oppia/blob/develop/core/domain/question_fetchers.py#L139). We will not be able to avoid this particular case, I think either the validation will be called for the question or it should not be called at all.
What should I do in this case?
Thanks

Reply:
Ah sorry, I missed this. OK, do the following:

1. In state_domain.State.from_dict(), add an optional param "validate=True". The default arg should be True. Only in the question fetcher should you pass in validate=False explicitly and leave a TODO explaining why we're doing this.
2. That TODO should point to an issue that says something like "validate all question states by writing a migration + audit 
job that migrates all questions to their latest version and running it, then once everything passes, remove this TODO".

Does that work?

Reply:
Yep!
That works, thanks for the response.